app/modules/zoo-api/zooapi.factory.js

```diff
@@ -4,10 +4,20 @@ require('./zooapi.module.js')
     .factory('zooAPI', zooAPI);
 
 var ApiClient = require('panoptes-client/lib/api-client');
-
+var oauth = require('panoptes-client/lib/oauth')
 // @ngInject
-function zooAPI(zooAPIConfig) {
-
+function zooAPI(AnnotationsFactory, $location) {
+    ApiClient.beforeEveryRequest = function() {
+        return oauth.checkBearerToken()
+            .then(function (token) {
+                console.log('Token refreshed: ', token);
+            })
+            .catch(function (error) {
```

Comment:
`checkBearerToken()` doesn't throw an error if it can't refresh an expired token. It resolves to null instead.

Reply:
I don't know what's best practice with promises, but couldn't we [reject the promise](https://github.com/zooniverse/panoptes-javascript-client/blob/f26d810767ce210bc8b773efab431f46991a8d72/lib/oauth.js#L183) and throw an error when it resolves to `null`?

Reply:
Won't that break zooniverse/anti-slavery-manuscripts#251?

Reply:
the way, that function does reject with an error if no token is found. https://github.com/zooniverse/panoptes-javascript-client/blob/f26d810767ce210bc8b773efab431f46991a8d72/lib/oauth.js#L204-L208

Reply:
Errors are caught, and null is returned, here: https://github.com/zooniverse/panoptes-javascript-client/blob/595e694c71dd877ddbae4a13e670fc466314258e/lib/oauth.js#L279-L284

Reply:
re: _When a login token could not be refreshed when it's supposed to, should it reject a promise or return null (as it does in PJC 2.9.2 onwards)?_ I'm not sure what the best practice is either, since I haven't worked with many other login/session libraries - but whatever it is, we'll need to document it clearly.

Thoughts:
- Rejecting Promises has been one of those fiddly things that I've run into, and judging from the weirdness I encounter it's worth opening a separate discussion (Slack, yo?) to discuss expectations on how they're supposed to work.
- If the best practice _is_ to reject a promise, don't worry about ASM PR251 - that PR's locked to using PJC 2.9.2 and I can write a separate PR to future-proof ASM once PJC 2.9.5 or etc is released.

Reply:
@eatyourgreens do you know why `null` is returned? As I said, why not `reject(error)` here https://github.com/zooniverse/panoptes-javascript-client/blob/595e694c71dd877ddbae4a13e670fc466314258e/lib/oauth.js#L283?

Reply:
Because that's a catch block, so the original promise has already rejected at this point. Have a look at the sections about [chaining promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#Chaining), and [handling errors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#Error_propagation), at MDN.

The return value of the `catch` is passed to the next `then` block in the chain.

Reply:
This discussion seems to be diverting away from SW, and more into how the client should work. Would it be better off having that discussion on zooniverse/panoptes-javascript-client?

Reply:
As to why `null` is returned, have a look at https://github.com/zooniverse/panoptes-javascript-client/pull/82 @shaunanoordin found that the client was returning `undefined` for the token in some cases.